From 8cfe63cc777f6f9aa3fd48d51308ab4bb904a841 Mon Sep 17 00:00:00 2001
From: ZhangPeng <zhangpeng362@huawei.com>
Date: Thu, 6 Jul 2023 19:22:18 +0800
Subject: [PATCH 630/780] [Huawei] userswap: fix LRU list_del corruption

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188960

--------------------------------

When executing high-concurrency test cases of uswap, we encountered a
list_del corruption as follows.

list_del corruption. next->prev should be fffffe2f8f3c9148, but was
ffff800052bbb8e8
WARNING: CPU: 96 PID: 4178 at lib/list_debug.c:62
__list_del_entry_valid+0x224/0x240
pc : __list_del_entry_valid+0x224/0x240
lr : __list_del_entry_valid+0x218/0x240
Call trace:
__list_del_entry_valid+0x224/0x240
release_pages+0x8fc/0x1300
free_pages_and_swap_cache+0x250/0x450
tlb_flush_mmu+0x360/0xd0c
zap_pte_range+0xfc0/0x1de0
unmap_page_range+0x3cc/0x1250
unmap_single_vma+0xb4/0x1f0
unmap_vmas+0x94/0x164
exit_mmap+0x12c/0x434
mmput+0x10c/0x540
exit_mm+0x4bc/0x7b4
do_exit+0x2f0/0xa7c

The LRU pagevecs might still hold a reference to the page if we didn't
drain them. In this case, remapping and then unmapping will corrupt the
LRU list that still holds the reference.
Fix this by adding page_count() check for swap-in VA to make sure that
no other kernel mechanism is using this physical page. Add
lru_add_drain_all() before swap-in VA to correct page_count() of swap-in
pages.

Signed-off-by: ZhangPeng <zhangpeng362@huawei.com>
---
 mm/userswap.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/mm/userswap.c b/mm/userswap.c
index f0ae208c3bdb..286260ea25be 100644
--- a/mm/userswap.c
+++ b/mm/userswap.c
@@ -433,6 +433,13 @@ int mfill_atomic_pte_nocopy(struct mm_struct *mm,
 		ret = -ENXIO;
 		goto out_put_page;
 	}
+	if (!PageLRU(page))
+		lru_add_drain_all();
+	if (page_mapcount(page) > 1 ||
+	    page_mapcount(page) + 1 != page_count(page)) {
+		ret = -EBUSY;
+		goto out_put_page;
+	}
 	ret = uswap_unmap_anon_page(mm, src_vma, src_addr, page, src_pmd,
 				    &src_pte, false);
 	if (ret)
-- 
2.17.1

